File: web/src/types/problem.ts
export type AssignedUser = {
  id: number
  name: string
  role: string
}

export type AssignedTeam = {
  id: number
  name: string
}

export type Attempt = {
  attempt_status: string | null
  attempt_phase: string | null
  problem_faced: string | null
  time_spent_min: number | null
  notes: string | null
  updated_at: string | null
} | null

export type TeamMemberAttempt = {
  id: number
  name: string
  role_in_team: "Member" | "Reserve"
  attempt_status: string | null
  attempt_phase: string | null
  problem_faced: string | null
  time_spent_min: number | null
  notes: string | null
  updated_at: string | null
}

export type TeamSummary = {
  team_id: number | null
  team_name: string
  solved: number
  total: number
  reserve_solved: number
  reserve_total: number
  members: TeamMemberAttempt[]
}

export type Problem = {
  id: number
  name: string
  url: string
  platform: string | null
  contest_type: string | null
  rating: number | null
  difficulty: string | null
  topic: string | null
  sub_topic: string | null
  tags: string[]
  importance: string | null
  key_idea: string | null
  status: string | null
  created_by: number | null
  notes: string | null
  date_added: string | null
  date_updated: string | null
  assigned_users: AssignedUser[]
  assigned_teams: AssignedTeam[]
  my_attempt: Attempt
  team_summary: TeamSummary[]
  /** Contestant view: who assigned this to the viewer, and how. */
  assigned_by: AssignedUser | null
  assigned_via: string | null
}

export type ProblemListResponse = {
  total: number
  count: number
  results: Problem[]
}

export type ProblemFilters = {
  q?: string
  // Multi-select capable: the Problem Bank passes arrays; the Assigned
  // sidebar passes single strings. The API client serialises both.
  platform?: string | string[]
  topic?: string | string[]
  difficulty?: string | string[]
  importance?: string | string[]
  status?: string | string[]
  contest_type?: string | string[]
  assigned_user_id?: number
  assigned_team_id?: number
  rating_min?: number
  rating_max?: number
  sort?: string
  order?: "asc" | "desc"
  limit?: number
  offset?: number
}

File: backend/app/schemas/problem.py
"""
Problem schemas.

Tags storage note: the `problems.tags` column is `text` holding a JSON-encoded
array (legacy convention).  The `Problem.tags` field below uses a
BeforeValidator to deserialize that string into a Python list, so callers
always see `list[str]`.
"""

from __future__ import annotations

import json
from datetime import datetime
from typing import Annotated, Any, Literal

from pydantic import BaseModel, BeforeValidator, ConfigDict, Field

from app.schemas.user import BoolFromInt


# ---------------------------------------------------------------------------
# Tag (de)serialization
# ---------------------------------------------------------------------------

def _parse_tags(value: object) -> list[str]:
    """Accept str (JSON-encoded) | list[str] | None and normalise to list[str]."""
    if value is None or value == "":
        return []
    if isinstance(value, list):
        return [str(x).strip() for x in value if str(x).strip()]
    if isinstance(value, str):
        try:
            parsed = json.loads(value)
            if isinstance(parsed, list):
                return [str(x).strip() for x in parsed if str(x).strip()]
        except json.JSONDecodeError:
            return [t.strip() for t in value.split(",") if t.strip()]
    return []


TagList = Annotated[list[str], BeforeValidator(_parse_tags)]


# ---------------------------------------------------------------------------
# Nested rows on a problem (assignment + per-viewer attempt)
# ---------------------------------------------------------------------------

class AssignedUserRef(BaseModel):
    id: int
    name: str
    role: str


class AssignedTeamRef(BaseModel):
    id: int
    name: str


class AttemptInline(BaseModel):
    """The viewer's own attempt — `my_attempt` field on the problem row."""

    attempt_status: str | None = None
    attempt_phase: str | None = None
    problem_faced: str | None = None
    time_spent_min: int | None = None
    notes: str | None = None
    updated_at: datetime | None = None


class TeamSummaryMember(BaseModel):
    id: int
    name: str
    role_in_team: str | None = None  # 'Member'/'Reserve', None for "direct user" rows
    attempt_status: str | None = None
    problem_faced: str | None = None
    time_spent_min: int | None = None
    notes: str | None = None
    updated_at: datetime | None = None


class TeamSummary(BaseModel):
    """Coach/Admin view — per-team breakdown rendered in the expand row."""

    team_id: int | None = None  # None when this group is "Direct user assignments"
    team_name: str
    solved: int = 0
    total: int = 0
    reserve_solved: int = 0
    reserve_total: int = 0
    members: list[TeamSummaryMember] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Response
# ---------------------------------------------------------------------------

class ProblemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: int
    name: str
    url: str

    platform: str | None = None
    contest_name: str | None = None
    contest_type: str | None = None
    contest_year: int | None = None
    problem_index: str | None = None

    rating: int | None = None
    difficulty: str | None = None
    topic: str | None = None
    sub_topic: str | None = None
    tags: TagList = Field(default_factory=list)

    importance: str | None = None
    suggested_role: str | None = None
    prerequisites: str | None = None
    key_idea: str | None = None
    editorial_url: str | None = None
    time_limit_ms: int | None = None
    memory_limit_mb: int | None = None

    status: str | None = "Todo"
    assigned_to: str | None = None
    created_by: int | None = None
    is_bank: BoolFromInt = False
    from_contest: BoolFromInt = False
    notes: str | None = None

    date_added: datetime | None = None
    date_updated: datetime | None = None

    # Hydrated by repository
    assigned_users: list[AssignedUserRef] = Field(default_factory=list)
    assigned_teams: list[AssignedTeamRef] = Field(default_factory=list)
    my_attempt: AttemptInline | None = None
    team_summary: list[TeamSummary] = Field(default_factory=list)


class ProblemListResponse(BaseModel):
    total: int
    count: int
    results: list[ProblemOut]


class ProblemDeleted(BaseModel):
    deleted: int


# ---------------------------------------------------------------------------
# Write bodies
# ---------------------------------------------------------------------------

class ProblemWrite(BaseModel):
    """Common write shape — used by POST (required url+platform enforced in
    router) and PUT/PATCH.  Fields not provided are left untouched on update."""

    model_config = ConfigDict(extra="ignore")

    name: str | None = None
    url: str | None = None
    platform: str | None = None
    contest_name: str | None = None
    contest_type: str | None = None
    contest_year: int | None = None
    problem_index: str | None = None
    rating: int | None = None
    difficulty: str | None = None
    topic: str | None = None
    sub_topic: str | None = None
    # Frontend may send list OR comma-separated string; tolerant on the way in.
    tags: list[str] | str | None = None
    importance: str | None = None
    suggested_role: str | None = None
    prerequisites: str | None = None
    key_idea: str | None = None
    editorial_url: str | None = None
    time_limit_ms: int | None = None
    memory_limit_mb: int | None = None
    status: str | None = None
    assigned_to: str | None = None
    notes: str | None = None

    # Many-to-many assignments — if omitted, the existing list is preserved.
    # If passed (even as []), it REPLACES the existing list.
    assigned_user_ids: list[int] | None = None
    assigned_team_ids: list[int] | None = None


class BulkCreateIn(BaseModel):
    items: list[dict[str, Any]] = Field(default_factory=list)


class BulkCreateResult(BaseModel):
    inserted: int
    skipped: int
    errors: list[dict[str, Any]] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Query params for GET /api/problems and /api/bank/problems
# ---------------------------------------------------------------------------

SortKey = Literal["name", "rating", "date_added", "importance", "id", "difficulty"]
SortOrder = Literal["asc", "desc"]


class ProblemFilters(BaseModel):
    """All query params the list endpoint accepts."""

    q: str | None = None
    platform: str | None = None
    topic: str | None = None
    difficulty: str | None = None
    importance: str | None = None
    status: str | None = None
    suggested_role: str | None = None
    contest_type: str | None = None
    contest_year: int | None = None
    rating_min: int | None = None
    rating_max: int | None = None
    sort: SortKey = "date_added"
    order: SortOrder = "desc"
    limit: int | None = Field(default=None, ge=1, le=500)
    offset: int = Field(default=0, ge=0)
